src/evo/providertx.cpp

```diff
@@ -0,0 +1,236 @@
+// Copyright (c) 2018-2021 The Dash Core developers
+// Copyright (c) 2021 The PIVX Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include "evo/providertx.h"
+
+#include "base58.h"
+#include "core_io.h"
+#include "masternode.h"     // MN_COLL_AMT
+#include "messagesigner.h"
+#include "evo/specialtx.h"
+#include "tinyformat.h"
+#include "utilstrencodings.h"
+#include "validation.h"
+
+/* -- Helper static functions -- */
+
+static bool CheckService(const CService& addr, CValidationState& state)
+{
+    if (!addr.IsValid()) {
+        return state.DoS(10, false, REJECT_INVALID, "bad-protx-ipaddr");
+    }
+    if (!Params().IsRegTestNet() && !addr.IsRoutable()) {
+        return state.DoS(10, false, REJECT_INVALID, "bad-protx-ipaddr");
+    }
+
+    // IP port must be the default one on main-net, which cannot be used on other nets.
+    static int mainnetDefaultPort = CreateChainParams(CBaseChainParams::MAIN)->GetDefaultPort();
+    if (Params().NetworkIDString() == CBaseChainParams::MAIN) {
+        if (addr.GetPort() != mainnetDefaultPort) {
+            return state.DoS(10, false, REJECT_INVALID, "bad-protx-ipaddr-port");
+        }
+    } else if (addr.GetPort() == mainnetDefaultPort) {
+        return state.DoS(10, false, REJECT_INVALID, "bad-protx-ipaddr-port");
+    }
+
+    // !TODO: add support for IPv6 and Tor
+    if (!addr.IsIPv4()) {
+        return state.DoS(10, false, REJECT_INVALID, "bad-protx-ipaddr");
+    }
+
+    return true;
+}
+
+template <typename Payload>
+static bool CheckStringSig(const Payload& pl, const CKeyID& keyID, CValidationState& state)
+{
+    std::string strError;
+    if (!CMessageSigner::VerifyMessage(keyID, pl.vchSig, pl.MakeSignString(), strError)) {
+        return state.DoS(100, false, REJECT_INVALID, "bad-protx-sig", false, strError);
+    }
+    return true;
+}
+
+template <typename Payload>
+static bool CheckInputsHash(const CTransaction& tx, const Payload& pl, CValidationState& state)
```

Comment:
This function is not used anywhere.

Would be good to add it to the unit test cases, validate that it throws the "bad-protx-inputs-hash" error.

Reply:
Yes, nice catch!
This was included the original commit (https://github.com/PIVX-Project/PIVX/commit/db4007ecafd3b7bc5484d989aea6586b6ca5f5fb) but was missed after #2270, in the refactoring to split contextual/non-contextual checks for special txes.
Added it back, and updated the tests as well (good idea).